src/listpack.c

```diff
@@ -941,3 +1039,806 @@ int lpValidateIntegrity(unsigned char *lp, size_t size, int deep){
 
     return 1;
 }
+
+unsigned int lpCompare(unsigned char *p, unsigned char *s, uint32_t slen) {
+    unsigned char buf[LP_INTBUF_SIZE];
+    unsigned char *value;
+    int64_t sz;
+
+    if (p[0] == LP_EOF) return 0;
+    value = lpGet(p, &sz, buf);
+    return (slen == sz) && memcmp(value,s,slen) == 0;
+}
+
+/* uint compare for qsort */
+static int uintCompare(const void *a, const void *b) {
+    return (*(unsigned int *) a - *(unsigned int *) b);
+}
+
+/* Helper method to store a string into from val or lval into dest */
+static inline void lpSaveValue(unsigned char *val, unsigned int len, int64_t lval, listpackEntry *dest) {
+    dest->sval = val;
+    dest->slen = len;
+    dest->lval = lval;
+}
+
+void lpRandomPair(unsigned char *lp, unsigned long total_count, listpackEntry *key, listpackEntry *val) {
+    unsigned char *p;
+    int64_t vlen;
+
+    /* Avoid div by zero on corrupt listpack */
+    assert(total_count);
+
+    /* Generate even numbers, because listpack saved K-V pair */
+    int r = (rand() % total_count) * 2;
+    p = lpSeek(lp, r);
+    assert(p);
+    key->sval = lpGet(p, &vlen, NULL);
+    if (key->sval) {
+        key->slen = vlen;
+    } else {
+        key->lval = vlen;
+    }
+
+    if (!val)
+        return;
+    p = lpNext(lp, p);
+    assert(p);
+    val->sval = lpGet(p, &vlen, NULL);
+    if (val->sval) {
+        val->slen = vlen;
+    } else {
+        val->lval = vlen;
+    }
+}
+
+void lpRandomPairs(unsigned char *lp, unsigned int count, listpackEntry *keys, listpackEntry *vals) {
+    unsigned char *p, *key, *value;
+    unsigned int klen = 0, vlen = 0;
+    int64_t klval = 0, vlval = 0, ele_len = 0;
+
+    /* Notice: the index member must be first due to the use in uintCompare */
+    typedef struct {
+        unsigned int index;
+        unsigned int order;
+    } rand_pick;
+    rand_pick *picks = zmalloc(sizeof(rand_pick)*count);
+    unsigned int total_size = lpLength(lp)/2;
+
+    /* Avoid div by zero on corrupt listpack */
+    assert(total_size);
+
+    /* create a pool of random indexes (some may be duplicate). */
+    for (unsigned int i = 0; i < count; i++) {
+        picks[i].index = (rand() % total_size) * 2; /* Generate even indexes */
+        /* keep track of the order we picked them */
+        picks[i].order = i;
+    }
+
+    /* sort by indexes. */
+    qsort(picks, count, sizeof(rand_pick), uintCompare);
+
+    /* fetch the elements form the listpack into a output array respecting the original order. */
+    unsigned int lpindex = picks[0].index, pickindex = 0;
+    p = lpSeek(lp, lpindex);
+    while (p && pickindex < count) {
+        key = lpGet(p, &ele_len, NULL);
+        if (key) {
+            klen = ele_len;
+        } else {
+            klval = ele_len;
+        }
+        p = lpNext(lp, p);
+        assert(p);
```

Comment:
```suggestion
        assert((p = lpNext(lp, p)));
```
if we take it, we can do the same on the two assertions in the next function too

Reply:
Good.